### PR TITLE
bpo-38610: Fix possible crashes in several list methods

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Hold strong references to list elements while calling
PyObject_RichCompareBool().

https://bugs.python.org/issue38610

### CODE REVIEW

File: Objects/listobject.c [hunk 1]
```diff
@@ -2580,7 +2583,10 @@ list_count(PyListObject *self, PyObject *value)
     Py_ssize_t i;
 
     for (i = 0; i < Py_SIZE(self); i++) {
-        int cmp = PyObject_RichCompareBool(self->ob_item[i], value, Py_EQ);
+        PyObject *obj = self->ob_item[i];
+        Py_INCREF(obj);
+        int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
```
Comment (shihai1991) [inline Objects/listobject.c:2588]: How about add the refcount in PyObject_RichCompareBool()?

File: Lib/test/test_list.py [hunk 2]
```diff
@@ -171,5 +171,30 @@ def test_preallocation(self):
         self.assertEqual(iter_size, sys.getsizeof(list([0] * 10)))
         self.assertEqual(iter_size, sys.getsizeof(list(range(10))))
 
+    def test_count_index_remove_crashes(self):
+        # The count(), index(), and remove() methods were not holding strong
```
Comment (pablogsal) [inline Lib/test/test_list.py]: I would suggest to include a reference to the issue in the comment as this is a regression test.

**Review by pablogsal — APPROVED** [review]
LGTM

File: Objects/listobject.c (hunk 1 above)
Comment (pablogsal) [reply Objects/listobject.c]: Managing the reference count is the responsibility of the caller, not of `PyObject_RichCompareBool`.

File: Lib/test/test_list.py (hunk 2 above)
Comment (ZackerySpytz) [reply Lib/test/test_list.py]: I've added the BPO number to the comment.

**Comment by pablogsal** [comment]
Thanks @ZackerySpytz for the quick fix!

**Pull request merged**